src/Microsoft.ML.Core/Data/ColumnType.cs

```diff
@@ -478,42 +478,39 @@ public override bool Equals(ColumnType other)
     /// class numbers, etc. For example, in multi-class classification, the label is typically
     /// a class number which is naturally a KeyType.
     ///
-    /// KeyTypes can be contiguous (the class number example), in which case they can have
-    /// a cardinality/Count. For non-contiguous KeyTypes the Count property returns zero.
-    /// Any KeyType (contiguous or not) can have a Min value. The Min value is always >= 0.
+    /// KeyTypes have a cardinality/Count that is strictly positive.
     ///
     /// Note that the representation value does not necessarily match the logical value.
-    /// For example, if a KeyType has range 1000-5000, then it has a Min of 1000, Count
-    /// of 4001, but the representational values are 1-4001. The representation value zero
-    /// is reserved to mean none/invalid.
+    /// For example, if a KeyType has range 0-5000, then it has a Count of 5001, but
+    /// the representational values are 1-5001. The representation value zero is reserved
+    /// to mean none/invalid.
     /// </summary>
     public sealed class KeyType : PrimitiveType
     {
-        private KeyType(Type type, DataKind kind, ulong min, int count, bool contiguous)
+        private KeyType(Type type, DataKind kind, ulong count)
             : base(type, kind)
         {
             Contracts.AssertValue(type);
             Contracts.Assert(kind.ToType() == type);
-
-            Contracts.CheckParam(min >= 0, nameof(min));
-            Contracts.CheckParam(count >= 0, nameof(count), "Must be non-negative.");
-            Contracts.CheckParam((ulong)count <= ulong.MaxValue - min, nameof(count));
-            Contracts.CheckParam((ulong)count <= kind.ToMaxInt(), nameof(count));
-            Contracts.CheckParam(contiguous || count == 0, nameof(count), "Must be 0 for non-contiguous");
-
-            Contiguous = contiguous;
-            Min = min;
+            Contracts.CheckParam(0 < count && count <= kind.ToMaxInt(), nameof(count),
```

Comment:
><= [](start = 52, length = 2)

Shouldn't this inequality be strict?

Reply:
I think that it should not be strict. What `Count` expresses is the **largest** possible value for a key. The bottleneck is the key underlying `type`. At most a key can take the largest representable value in that type, which is kind.ToMaxInt().  

---
In reply to: [249202365](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249202365) [](ancestors = 249202365)

Reply:
Right, so, if I have a `U1`, values 1 through 255 would represent the range of possible keys, and correspond to a count of 255, which equals I think `kind.ToMaxInt`. So I think as written it is correct. Though I agree it is slightly confusing and you have to think about it a while before realizing it is correct -- something about it somehow just tickles my brain every time I see it, not quite sure what.

---
In reply to: [249229341](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249229341) [](ancestors = 249229341,249202365)